### PR TITLE
Added a huge pile of python utilities

### PR DESCRIPTION
**Pull request by tpoterba**
Includes:
 - Type system in python that mirrors scala
 - Annotation conversion system, and full support for all annotation objects in python
 - query_variants and query_samples functions on VDS that return python objects
 - implemented Variant, Genotype, etc as first class python objects and fully documented them
 - tests for the above

### CODE REVIEW

File: python/hail/dataset.py [hunk 1]
```diff
@@ -235,33 +264,22 @@ def annotate_alleles_expr(self, condition, propagate_gq=False):
             pargs.append('--propagate-gq')
         return self.hc.run_command(self, pargs)
 
-    def annotate_global_expr_by_variant(self, condition):
-        """Update the global annotations with expression with aggregation over
-        variants.
+    def annotate_global_py(self, annotation, ptype, path):
```
Comment (cseed) [inline python/hail/dataset.py]: Don't call it ptype, call it type.  In the python API, all X are python X.

Comment (cseed) [inline python/hail/dataset.py]: I'd probably put the path first in analog with `va.info.callStats = gs.callStats()`, etc.

File: python/hail/dataset.py [hunk 2]
```diff
@@ -27,6 +38,24 @@ def sample_ids(self):
...
+        :return: dict
+        """
+
+        schema = Type._from_java(self.jvds.saSignature())
+        zipped_annotations = scala_package_object(self.hc.hail.utils).iterableToArrayList(
+            self.jvds.sampleIdsAndAnnotations()
+        )
+        r = {}
+        for element in zipped_annotations:
+            r[element._1()] = schema.convert_to_py(element._2())
```
Comment (cseed) [inline python/hail/dataset.py]: Make convet_to_py private, too.

File: python/hail/dataset.py [hunk 3]
```diff
@@ -2422,35 +2450,184 @@ def persist(self, storage_level="MEMORY_AND_DISK"):
...
+    def global_schema(self):
+        return Type._from_java(self.jvds.globalSignature())
 
-        :param output: Output file.
-        :type output: str or None
+    def sample_schema(self):
+        return Type._from_java(self.jvds.saSignature())
 
-        :param bool attributes: If True, print attributes.
+    def variant_schema(self):
```
Comment (cseed) [inline python/hail/dataset.py]: Short documentation for these three?

File: python/hail/dataset.py [hunk 4]
```diff
@@ -2422,35 +2450,184 @@ def persist(self, storage_level="MEMORY_AND_DISK"):
...
+        return Type._from_java(self.jvds.vaSignature())
 
-        :param bool va: If True, print variant annotations schema.
+    def query_samples_typed(self, exprs):
```
Comment (cseed) [inline python/hail/dataset.py]: Why is this called _typed?  The examples are just query_samples.

Comment (cseed) [inline python/hail/dataset.py]: Ah, I see.  I'd put the main documentation on query_samples and just a short description on _typed versions about the difference.

File: python/hail/dataset.py [hunk 5]
```diff
@@ -2541,12 +2718,12 @@ def sample_qc(self, branching_factor=None):
 
     def show_globals(self, output=None):
         """Print or export all global annotations as JSON
-
         :param output: Output file.
         :type output: str or None
-
         """
 
+        warnings.warn('This function is deprecated.  Use the .globals() method instead.', DeprecationWarning)
```
Comment (cseed) [inline python/hail/dataset.py]: Kill it!

File: python/hail/java.py [hunk 6]
```diff
@@ -1,21 +1,50 @@
...
+    jarr = Env.gateway().new_array(jtype, len(lst))
     for i, s in enumerate(lst):
         jarr[i] = s
     return jarr
 
+
 def scala_object(jpackage, name):
     return getattr(getattr(jpackage, name + '$'), 'MODULE$')
 
+
```
Comment (cseed) [inline python/hail/java.py]: Any particular reason for the extra new line?

File: python/hail/java.py [hunk 7]
```diff
@@ -1,21 +1,50 @@
...
+def jsome(x):
+    return Env.jvm().scala.Some(x)
+
+
+def joption(x):
+    return jsome(x) if x else jnone()
+
+
+def strip_option(x):
```
Comment (cseed) [inline python/hail/java.py]: Consider from_option?

File: python/hail/representation/genotype.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    """
+    An object that represents an individual's genotype at a genomic locus.
+
+    :param int gt: Genotype hard call
+    :param ad: allelic depth (1 element per allele including reference)
+    :type ad: list of int
+    :param int dp: total depth
+    :param int gq: genotype quality
+    :param pl: phred-scaled posterior genotype likelihoods (1 element per possible genotype)
+    :type pl: list of int
```
Comment (cseed) [inline python/hail/representation/genotype.py]: types can also be None.  Should gt also default to None?

File: python/hail/representation/genotype.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        if ad:
+            ad = jsome(jarray(jvm.int, ad))
+        else:
+            ad = jnone()
+        dp = joption(dp)
+        gq = joption(gq)
+        if pl:
+            pl = jsome(jarray(jvm.int, pl))
+        else:
+            pl = jnone()
```
Comment (cseed) [inline python/hail/representation/genotype.py]: I would have left the attributes as python types.  Also, you don't want to make them attributes?  You're afraid someone will modify them?  Getter-only property is another option:

http://stackoverflow.com/questions/14594120/python-read-only-property

**Comment by cseed** [comment]
Also, this is totally fucking awesome.  Great work, @tpoterba.

File: python/hail/java.py (hunk 6 above)
Comment (tpoterba) [inline python/hail/java.py]: module-level functions are treated like classes according to my intelliJ python-style.  I've mostly just been autoformatting as I go, and assuming that's what we want.

File: python/hail/representation/genotype.py (hunk 8 above)
Comment (tpoterba) [inline python/hail/representation/genotype.py]: I like requiring gt -- can always set to none explicitly.

File: python/hail/java.py (hunk 7 above)
Comment (tpoterba) [inline python/hail/java.py]: Agree

File: python/hail/dataset.py (hunk 5 above)
Comment (tpoterba) [inline python/hail/dataset.py]: Done!

File: python/hail/representation/genotype.py (hunk 9 above)
Comment (tpoterba) [inline python/hail/representation/genotype.py]: this is a great idea, I've done this for the other representation types too.

**Comment by tpoterba** [comment]
Ready for a look.

File: python/hail/dataset.py [hunk 10]
```diff
@@ -1586,39 +1558,42 @@ def filter_genotypes(self, condition, keep=True):
...
-        return self.hc.run_command(self, pargs)
+        return self.hc._run_command(self, pargs)
 
     def filter_samples_all(self):
-        """Removes all samples from VDS.  The variants and variant annotations will
-        remain, making it a sites-only VDS.
+        """Removes all samples from VDS.
 
-        :return: A sites-only Variant Dataset
+        The variants and variant annotations will remain, making it a sites-only VDS.
```
Comment (cseed) [inline python/hail/dataset.py]: Mention global annotations remain as you do below.

File: python/hail/dataset.py [hunk 11]
```diff
@@ -1811,28 +1788,31 @@ def filter_variants_list(self, input, keep=True):
...
         pargs = ['filtervariants', 'list',
                  '--keep' if keep else '--remove',
                  '-i', input]
-        return self.hc.run_command(self, pargs)
+        return self.hc._run_command(self, pargs)
+
+    def globals(self):
+        """Return global annotations as a python object.
+
+        :return: vds global annotations
```
Comment (cseed) [inline python/hail/dataset.py]: I'd say "Global annotations." or "Dataset global annotations."   VDS sounds like a file (and should be capitalized).

File: python/hail/dataset.py [hunk 12]
```diff
@@ -1811,28 +1788,31 @@ def filter_variants_list(self, input, keep=True):
...
+        """
+
+        return Type._from_java(self._jvds.globalSignature())._convert_to_py(self._jvds.globalAnnotation())
```
Comment (cseed) [inline python/hail/dataset.py]: Why not use `global_schema`?

File: python/hail/dataset.py [hunk 13]
```diff
@@ -2386,10 +2368,10 @@ def pca(self, scores, loadings=None, eigenvalues=None, k=10, as_array=False):
             pargs.append(eigenvalues)
         if as_array:
             pargs.append('--arrays')
-        return self.hc.run_command(self, pargs)
+        return self.hc._run_command(self, pargs)
 
     def persist(self, storage_level="MEMORY_AND_DISK"):
-        """Persist the current dataset.
+        """Persist the current dataset to memory and disk.
```
Comment (cseed) [inline python/hail/dataset.py]: Memory only and disk only are both options.  Memory and/or disk?

File: python/hail/dataset.py [hunk 14]
```diff
@@ -2410,47 +2392,173 @@ def persist(self, storage_level="MEMORY_AND_DISK"):
...
+
+        See :py:meth:`.query_samples` for more information.
+
+        :param exprs: one or more query expressions
+        :type exprs: str or list of str
+        :return: annotation(s) and type(s)
+        :rtype: tuple of annotation or list of annotation, and :class:`.Type` or list of :class:`.Type`
+        """
+
+        if (isinstance(exprs, list)):
```
Comment (cseed) [inline python/hail/dataset.py]: I'm confused by this.  Do you allow exprs to be a string with a comma-separated list of expresions?  Then I think this does the wrong thing.

I've never been a fan of this "if it is one you get the value, more you get a list."  I see the convenience generally it makes things more complicated, esp. if you want to use this in python code that builds pipelines programmatically (which everyone is doing).

File: python/hail/dataset.py [hunk 15]
```diff
@@ -2784,26 +2867,20 @@ def tdt(self, fam, root='va.tdt'):
...
         pargs = ['tdt', '--fam', fam, '--root', root]
-        return self.hc.run_command(self, pargs)
+        return self.hc._run_command(self, pargs)
 
     def typecheck(self):
-        """Check if all sample, variant and global annotations are consistent
-        with the schema.
-
-        """
+        """Check if all sample, variant and global annotations are consistent with the schema."""
```
Comment (cseed) [inline python/hail/dataset.py]: This is really for internal use.  I'm happy if you make it private, too.

File: python/hail/dataset.py [hunk 16]
```diff
@@ -3107,24 +3177,41 @@ def variants_keytable(self):
           Struct {
             v: Variant
             va: variant annotations
-	  }
+	      }
```
Comment (cseed) [inline python/hail/dataset.py]: Looks like spacing broke here.

File: python/hail/dataset.py [hunk 17]
```diff
@@ -3107,24 +3177,41 @@ def variants_keytable(self):
...
+        """Convert samples and sample annotations to KeyTable.
+
+        The resulting KeyTable has schema:
+
+        .. code-block:: text
+
+          Struct {
+            s: Sample
+            sa: sample annotations
+	       }
```
Comment (cseed) [inline python/hail/dataset.py]: Spacing.

File: python/hail/java.py (hunk 6 above)
Comment (cseed) [inline python/hail/java.py]: Yep, sounds good, thanks.

File: python/hail/keytable.py [hunk 18]
```diff
@@ -1,55 +1,50 @@
...
             return self.jkt.nFields()
         except Py4JJavaError as e:
-            self._raise_py4j_exception(e)
+            raise_py4j_exception(e)
 
     def schema(self):
         """KeyTable schema.
 
-        :rtype: :class:`.Type`
+        >>> print(kt.schema())
```
Comment (cseed) [inline python/hail/keytable.py]: Examples section, and/or (short) explanation?  Or is it obvious?

File: python/hail/keytable.py [hunk 19]
```diff
@@ -60,30 +55,27 @@ def key_names(self):
...
         """
         try:
             return self.jkt.nRows()
         except Py4JJavaError as e:
-            self._raise_py4j_exception(e)
-        
+            raise_py4j_exception(e)
 
     def same(self, other):
         """Test whether two KeyTables are identical.
```
Comment (cseed) [inline python/hail/keytable.py]: You changed Variant Dataset to dataset (which I totally agree with).  Should we keep writing KeyTable here?

Comment (cseed) [inline python/hail/keytable.py]: Ah, I see you did below.  Do this one, too.

File: python/hail/keytable.py [hunk 20]
```diff
@@ -276,12 +262,10 @@ def aggregate_by_key(self, key_condition, agg_condition):
 
         :param key_condition: Named expression(s) for how to compute the keys of the new KeyTable.
         :type key_condition: str or list of str
-
         :param agg_condition: Named aggregation expression(s).
         :type agg_condition: str or list of str
 
         :return: A new KeyTable with the keys computed from the ``key_condition`` and the remaining columns computed from the ``agg_condition``.
```
Comment (cseed) [inline python/hail/keytable.py]: A new key table ...

File: python/hail/representation/genotype.py [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+
+        This function returns None if the genotype call is missing.
+
+        :param int num_genotypes: number of possible genotypes
+        :rtype: list of int or None
+        """
+
+        result = from_option(self._jrep.oneHotGenotype(num_genotypes))
+        if result:
+            result = [x for x in scala_package_object(Env.hail_package().utils).iterableToArrayList(result)]
```
Comment (cseed) [inline python/hail/representation/genotype.py]: You might make iterableToArrayList into into jiterable_to_list convenience function that handles nulls/None in the java package.

File: python/hail/type.py [hunk 22]
```diff
@@ -1,12 +1,582 @@
...
+        return self.msg
+
 
 class Type(object):
-    """Type of values."""
-    
+    """
+    Hail type class used for annotations and expression language.
+
+    :param jtype: equivalent java type
```
Comment (cseed) [inline python/hail/type.py]: Again, I suggest we don't document constructors with java objects the user can't/shouldn't call.

File: python/hail/type.py [hunk 23]
```diff
@@ -1,12 +1,582 @@
...
+        return self._jtype.toPrettyString(False, False)
+
+    def __eq__(self, other):
+        return self._jtype.equals(other._jtype)
+
+    @classmethod
+    def _from_java(cls, jtype):
+        # FIXME this is pretty hacky
+        class_name = jtype.getClass().getCanonicalName()
+        if class_name == 'is.hail.expr.TInt$':
```
Comment (cseed) [inline python/hail/type.py]: Static dictionary lookup is probably faster for singleton types.

File: python/hail/type.py [hunk 24]
```diff
@@ -1,12 +1,582 @@
...
+        elif class_name == 'is.hail.expr.TLocus$':
+            return TLocus()
+        elif class_name == 'is.hail.expr.TInterval$':
+            return TInterval()
+        elif class_name == 'is.hail.expr.TGenotype$':
+            return TGenotype()
+        else:
+            raise TypeError("unknown type class: '%s'" % class_name)
+
+    def _jtype(self):
```
Comment (cseed) [inline python/hail/type.py]: I think this is unpythonic.  Nuke 'em.

File: python/hail/type.py [hunk 25]
```diff
@@ -1,12 +1,582 @@
...
+        Raise an exception if the given annotation is not the appropriate type.
+
+        :param annotation: value to check
+        """
+        raise NotImplemented('use a subclass of Type')
+
+
+class TInt(Type):
+    """
+    Hail type corresponding to int
```
Comment (cseed) [inline python/hail/type.py]: You mean python int?  I don't think that's quite right.  The size of python ints are system dependent.  TInt is 32-bit integers.

File: python/hail/type.py [hunk 26]
```diff
@@ -1,12 +1,582 @@
...
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation and not isinstance(annotation, int):
+            raise TypeCheckError("TInt expected type 'int', but found type '%s'" % type(annotation))
+
+
+class TLong(Type):
+    """
+    Hail type corresponding to long
```
Comment (cseed) [inline python/hail/type.py]: Same.  64-bit integers.

File: python/hail/type.py [hunk 27]
```diff
@@ -1,12 +1,582 @@
...
+            return annotation
+
+    def _typecheck(self, annotation):
+        if annotation and not (isinstance(annotation, long) or isinstance(annotation, int)):
+            raise TypeCheckError("TLong expected type 'int' or 'long', but found type '%s'" % type(annotation))
+
+
+class TFloat(Type):
+    """
+    Hail type corresponding to float
```
Comment (cseed) [inline python/hail/type.py]: Again, I don't think the size of Python floats is specified.  TFloat is 32-bit floating-point number.

File: python/hail/type.py [hunk 28]
```diff
@@ -1,12 +1,582 @@
...
+        raise NotImplementedError('TFloat is currently unsupported in certain operations, use TDouble instead')
+
+    def _typecheck(self, annotation):
+        if annotation and not isinstance(annotation, float):
+            raise TypeCheckError("TDouble expected type 'float', but found type '%s'" % type(annotation))
+
+
+class TDouble(Type):
+    """
+    Hail type corresponding to float
```
Comment (cseed) [inline python/hail/type.py]: 64-bit floating point number.  Note, this is the meaning on the Hail side.  On the python side, they are represented by float, yes.  You might mention both.

File: python/hail/type.py [hunk 29]
```diff
@@ -1,12 +1,582 @@
...
+        if annotation:
+            if not isinstance(annotation, set):
+                raise TypeCheckError("TSet expected type 'set', but found type '%s'" % type(annotation))
+            for elt in annotation:
+                self.element_type._typecheck(elt)
+
+
+class TDict(Type):
+    """
+    Hail type corresponding to dict
```
Comment (cseed) [inline python/hail/type.py]: Python dicts can have any key type.  Mention that Hail values have/require string keys.

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  def iterableToArrayList[T](it: Iterable[T]): java.util.ArrayList[T] = {
+    val list = new java.util.ArrayList[T]()
+    for (elem <- it)
+      list.add(elem)
+    list
+  }
+
+  def arrayListToISeq[T](al: java.util.ArrayList[T]): IndexedSeq[T] = al.asScala.toIndexedSeq
```
Comment (cseed) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: This returns a Vector.  What about `al.toArray(new Array[T](al.length)): IndexedSeq[T]`?

**Comment by tpoterba** [comment]
Addressed comments.

**Comment by tpoterba** [comment]
oops, need to fix tutorial

File: src/main/scala/is/hail/utils/Py4jUtils.scala (hunk 30 above)
Comment (tpoterba) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: That requires a classtag, and when I asked for the classtag I got a couple of other odd compiler errors.  I'll leave it as indexed seq for now with a FIXME note

**Comment by tpoterba** [comment]
Good to go, I think.

**Comment by cseed** [comment]
Here we go!

**Pull request merged**